stan/math/prim/fun/mean.hpp

```diff
@@ -10,36 +10,18 @@ namespace math {
 
 /**
  * Returns the sample mean (i.e., average) of the coefficients
- * in the specified standard vector.
+ * in the specified std vector, vector, row vector, or matrix.
  *
- * @tparam T type of elements in the vector
- * @param v Specified vector.
- * @return Sample mean of vector coefficients.
- * @throws std::domain_error if the size of the vector is less
- * than 1.
- */
-template <typename T>
-inline return_type_t<T> mean(const std::vector<T>& v) {
-  check_nonzero_size("mean", "v", v);
-  Eigen::Map<const Eigen::Matrix<T, Eigen::Dynamic, 1>> m(&v[0], v.size());
-  return m.mean();
-}
-
-/**
- * Returns the sample mean (i.e., average) of the coefficients
- * in the specified vector, row vector, or matrix.
- *
- * @tparam T type of elements in the matrix
- * @tparam R number of rows, can be Eigen::Dynamic
- * @tparam C number of columns, can be Eigen::Dynamic
+ * @tparam T type of the matrix
  *
- * @param m Specified vector, row vector, or matrix.
- * @return Sample mean of vector coefficients.
+ * @param m Specified std vector, vector, row vector, or matrix.
+ * @return Sample mean of container coefficients.
  */
-template <typename T, int R, int C>
-inline return_type_t<T> mean(const Eigen::Matrix<T, R, C>& m) {
+template <typename T, require_container_t<T>* = nullptr>
+inline return_type_t<T> mean(const T& m) {
   check_nonzero_size("mean", "m", m);
-  return m.mean();
+  return apply_vector_unary<T>::reduce(m,
+                                       [](const auto& a) { return a.mean(); });
```

Comment:
[Q] is this not the kind of place we would want to forward along `m`?

Reply:
No. `apply_vector_unary<T>::reduce()` does not return `m` or a struct/object containing `m`. It also does not make a copy of `m` into local variable. So it can work with const ref as well.

Whenever a quistion is raised about whether perfect forwarding should be used for a certain argument try going trough what happens when a functions accepts the argument as const ref or as a forwarding ref. Do any copy or move constructor needs to be invoked when passed a lvalue/rvalue? If no copying of argument can be avoided by perfect forwarding than it is not needed.